src/google/protobuf/compiler/csharp/csharp_helpers.cc

```diff
@@ -240,6 +239,121 @@ std::string GetPropertyName(const FieldDescriptor* descriptor) {
   return property_name;
 }
 
+// Utility functions for all case handling. Not using ctype.h due to locale complications.
+// TODO: Use in other code in this file.
+bool IsLowerLetter(char c) {
+    return 'a' <= c && c <= 'z';
+}
+
+bool IsUpperLetter(char c) {
+    return 'A' <= c && c <= 'Z';
+}
+
+bool IsNumeric(char c) {
+  return '0' <= c && c <= '9';
+}
+
+bool IsAlphaNumeric(char c) {
+    return IsLowerLetter(c) || IsUpperLetter(c) || IsNumeric(c);
+}
+
+// Upper-cases any lower-case letter; non-lower-letters are preserved as they were
+char ToUpper(char c) {
+    return IsLowerLetter(c) ? (c + 'A' - 'a') : c;
+}
+
+// Lower-cases any upper-case letter; non-upper-letters are preserved as they were
+char ToLower(char c) {
+    return IsUpperLetter(c) ? (c + 'a' - 'A') : c;
+}
+
+// Convert a string which is expected to be SHOUTY_CASE (but may not be *precisely* shouty)
+// into a PascalCase string. Precise rules implemented:
+
+// Previous input character      Current character         Case
+// Any                           Non-alphanumeric          Skipped
+// None - first char of input    Alphanumeric              Upper
+// Non-letter (e.g. _ or 1)      Alphanumeric              Upper
+// Numeric                       Alphanumeric              Upper
+// Lower letter                  Alphanumeric              Same as current
+// Upper letter                  Alphanumeric              Lower
+std::string ShoutyToPascalCase(const std::string& input) {
+  string result;
+  // Simple way of implementing "always start with upper"
+  char previous = '_';
+  for (int i = 0; i < input.size(); i++) {
+    char current = input[i];
+    if (!IsAlphaNumeric(current)) {
+      previous = current;
+      continue;
+    }
+    if (!IsAlphaNumeric(previous)) {
+      result += ToUpper(current);
+    } else if (IsNumeric(previous)) {
+      result += ToUpper(current);
+    } else if (IsLowerLetter(previous)) {
+      result += current;
+    } else {
+      result += ToLower(current);
+    }
+    previous = current;
+  }
+  return result;
+}
+
+// Generate the name to use in a generated C# enum from the enum name, the enum value name, and the naming style.
+std::string GetEnumValueName(const std::string& enum_name, const std::string& enum_value_name, CSharpEnumNaming namingStyle) {
+  if (namingStyle == CSHARPENUMNAMING_PRESERVE) {
+    return enum_value_name;
+  }
+  std:string effective_value_name = enum_value_name;
```

Comment:
Copy `enum_value_name` into `effective_value_name` in an else block instead, avoid this copy when `namingStyle` is `CSHARPENUMNAMING_PASCAL_TRIM_PREFIX`.